### ISSUE TITLE
Pairwise postprocessing validation is very slow - possible optimizations

### ISSUE THREAD

**Issue by korotaS** [OML-Team/open-metric-learning#599]
Hi! I've been trying to train a Reranking pairwise model (using [this guide](https://open-metric-learning.readthedocs.io/en/latest/postprocessing/siamese_examples.html) and OML version 3.1.0) and it seems to train ok but the validation takes too much time - one epoch of training takes about 5 minutes and one validation cycle takes about 4 hours (the dataset is pretty big). I've used `cProfile` to profile a part of training and here are the top 100 slowest functions: [train_rerank_cprofile.txt](https://github.com/user-attachments/files/15864039/train_rerank_cprofile.txt). The whole training took about 63 hours (229k seconds), validation is every 10 epochs.

There are 2 things that I've noticed and they definitely can be optimized: 

First - look at line `...48866.403...oml/datasets/images.py:281(get_query_ids)` and below `...48784.334...images.py:284(get_gallery_ids)` (line numbers might not match because I made some other minor changes), so two methods `get_query_ids` and `get_gallery_ids` that are called in `oml/retrieval/postprocessors/pairwise.py:121-122` take 97k seconds. query_ids and gallery_ids don't change during training, so we can inside `ImageQueryGalleryLabeledDataset.__init__` calculate query_ids and gallery_ids once and save about 97k seconds during training in my case. 

Second - it is not obvious from the `cProfile` txt file but from ClearML logs it seemed interesting that PL validation step take about 40 minutes although validation step looks like this: 
https://github.com/OML-Team/open-metric-learning/blob/05842c84c980fa2c34510e3de176cd5c5dae9205/oml/lightning/modules/pairwise_postprocessing.py#L84-L86
We don't use images that are read from disc directly after this step (we use only embeddings), however we load images **again** in the PairwiseDataset:
https://github.com/OML-Team/open-metric-learning/blob/05842c84c980fa2c34510e3de176cd5c5dae9205/oml/datasets/pairs.py#L34-L35
I think we can pass some kind of `load_images` parameter to the dataset `__init__` method and load them only when needed - for example, in training dataset and in pairs dataset, but not in validation dataset (or as alternative - make a wrapper to `__getitem__` method and pass this parameter directly there), it can save 40 minutes in each validation in my case. 

There were some other issues during training but they aren't related to optimization so I will create another issue.

**Comment by AlekseySh** [comment]
Hey, @korotaS, thank you for you interest in OML.

A general note on post-processing: it's expected that validation takes longer than training "epoch":
* first, training "epoch" is not a full epoch if you use samplers like "BalanceSampler", because it iterates over all labels (not over all instances) during the epoch
* second, we have more inferences in validation and every single of them is slower because images are concatenated 
* thus, I recommend to perform X training cycles, then 1 validation cycle (that you already did with X = 10, but it may be even bigger)

**Comment by AlekseySh** [comment]
> get_query_ids, get_gallery_ids

agree, even without post-processing context, it would be a good optimization

but are you really sure that we spent 26h from 63h doing this operation? I'm very skeptical... You can do this optimization locally first and check if numbers are really like this.

**Comment by AlekseySh** [comment]
As for the second point. You are right. There is a duplicated image loading when using Lightning.

Basically, we need the first loader (built on `LabeledDataset`) in validation step only to deliver and accumulate embeddings. So, if we want to remove duplicated loading, **we should do it here**. The second loader (built on `PairsDataset`) is okay, because it indeed needs to load images.

I would say, there there a two root problems here.
* First, Lightning is not super flexible, so we need to provide it with some dataset and loader to perform its validation steps.
* As a result, we used `ImageDataset`. It's used to deliver embeddings from `extra_data` container. Ideally, we should use `EmbeddingsDataset` for this purpose (by the way there are some datasets like this in OML, but in the test folder: `tests/test_integrations/utils.py`), but even in this case we get another problem: visualization needs to be handled somehow. 

If you take a look on pure python [examples](https://open-metric-learning.readthedocs.io/en/latest/postprocessing/siamese_examples.html) you will see, that there is no duplicated image reading in the validation. (We create `val_dataset`, but we don't iterate over all `__getitems__` to build the original retrieval results).

**Comment by AlekseySh** [comment]
Ideally, we should use `EmbeddingsQueryGalleryLabeledDataset` for the purpose of delivering and accumulating embeddings. We can move an existing datasets from `tests/test_integrations/utils.py` to the library core and update it a little bit.

The update is needed in order not to loose visalisation functionality which happens in validation. (When we used image dataset we had `visualise()` method). We can do it something like this:

```

def vis(i):
    return image_dataset.visualise(i)  # this dataset is used only to surve for visualisation, there is no loop over __getitem__

class EmbeddingsQueryGalleryLabeledDataset:

    def __init__ (..., visualisation_fn: Optional = vis):
        self.vis = vis


    def visualise(self, i):
        return self.vis[i]

```

**Comment by AlekseySh** [comment]
@korotaS got it, what is the domain? retail?

>  I will locally make this optimization and after that check whether the time of those methods decreased to ~0 seconds.

yep, it should be quick and informative 

see my other comments above and have a nice day :)

**Comment by korotaS** [comment]
The domain is indeed a retail. In fact, we were on the same meeting with Epoch8 last Tuesday and we were discussing this same problem after your presentation of new features of OML 😁

**Comment by AlekseySh** [comment]
I thought so, nice to meet you here!

Don't hesitate to join us, especially if you are going to contribute: https://t.me/+lqsKu2af8xcyMjEy

**Comment by korotaS** [comment]
As for your proposed solution about embeddings dataset - maybe I misunderstood something but I think that it won't work that way. If we create an `EmbeddingsQueryGalleryLabeledDataset` as a validation dataset, it will break at the `pairwise_inference` function because it won't know how to load an image. And the problem is that internally in `pairwise_inference` function the `PairDataset` is created as a wrapper for the validation dataset, so one depends on another. 

We can use `EmbeddingsQueryGalleryLabeledDataset` as a validation dataset, but at the same time `PairDataset` needs to be created on top of some dataset that **can** load images. Maybe we can instantiate an `ImageQueryGalleryLabeledDataset`, wrap it in `PairDataset`, pass it to `PairwiseReranker.__init__` and then pass this dataset to `pairwise_inference` function (however we will need to update `pair_ids` every time).

**Comment by AlekseySh** [comment]
>  If we create an EmbeddingsQueryGalleryLabeledDataset as a validation dataset, it will break at the pairwise_inference function because it won't know how to load an image.

Oh, you are right. So, the problem is that we want to use the same dataset first to deliver **embeddings**, second to be a base for `PairsDataset` to load pairs of **images**.

**Comment by AlekseySh** [comment]
I'm not sure that I understood you solution, but we can have another easy-to option to start with. Our datasets support cache on reading images (so, we cache bytes). Unfortunately, cache is not parsed in postprocessing config as it's done in the main training config. I can easily add it in a few lines of code. It's not a full solution (we still need some time to stack decode and stack images), but it should speed up the process a lot.

**Comment by AlekseySh** [comment]
As for the first issue, the solution is already in the main branch and will be released in pypi soon. Thank you for the report!

**Comment by korotaS** [comment]
>Could you please check how much it helps?

I've fetched your updates and set `cache_size` to a very big number (roughly equal to val dataset size), so that ~all of the `PairsDataset.__getitem__` calls in val loop are cached, but the validation loop itself (without `PairwiseReranker.process`) didn't speed up... Maybe the `cache_size` was too big and it can't work as expected with such bug number?

**Comment by korotaS** [comment]
Or maybe the bottleneck is not in reading images but in transforming them:
<img width="889" alt="image" src="https://github.com/OML-Team/open-metric-learning/assets/50669173/9df50c4b-c5e3-48eb-9c70-0ece328d0ee2">

As you can see, from 2726 seconds that `__getitem__` takes 2600 are taken by transforms

**Comment by AlekseySh** [comment]
Anyway, @korotaS , what percentage of speedup are we talking about if we avoid loading images the second time? (You can replace `__getitem__` method in a hacky way to understand it or use your `load_images` parameter)

**Comment by korotaS** [comment]
When I run validation as is, the embeddings accumulation part (before `pairwise_inference`) took about 40 minutes out of ~2h20m of full validation. When I used `load_images` and loaded images only in `PairsDataset`, 40 minutes decreased to 25 seconds, so the speedup is about 28% (40m / 140m). 

I will now try increasing cache_size to be bigger than val dataset size (if val dataset size is 100, than in embeddings accumulation 100 images will be loaded and then in `pairwise_inference` the same 100 images will be loaded, so we need cache_size to be bigger than val dataset size) and report on any speedup.

**Comment by AlekseySh** [comment]
@korotaS got it!
So, if we can save ~28% of validation and validation time takes 25% of the full training cycle (val + train), it means we can save 25*0.28~=7% of the full time. Not as much. Let's see if bigger cache helps.

The profit from that wrong ids handling was much higher)

Yep, in that case 101 cache size seems enough

**Comment by AlekseySh** [comment]
@korotaS what should we do with this issue? have you tried bigger cache size in order to speed up the process?

**Comment by korotaS** [comment]
I have tested the caching function, unfortunately I don't have the numbers now but it didn't show a good speedup. I think that is is because of transforms, as I said earlier: 

> Or maybe the bottleneck is not in reading images but in transforming them: <img alt="image" width="889" src="https://private-user-images.githubusercontent.com/50669173/340658537-9df50c4b-c5e3-48eb-9c70-0ece328d0ee2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwOTI4OTQsIm5iZiI6MTcyMDA5MjU5NCwicGF0aCI6Ii81MDY2OTE3My8zNDA2NTg1MzctOWRmNTBjNGItYzVlMy00OGViLTljNzAtMGVjZTMyOGQwZWUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDExMjk1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyYzdhMjIyZDZkZjBlYzhhMjQ5NjJiODY5MzMxNjJjMTkzNjhmOGJkNGVhNTZhZjJiMjAzMDVjY2YxZTg2N2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OS0W-qtHSkJShZ0yzWxUWP5u1S3OOzgJaKcO1xyonK8">
> 
> As you can see, from 2726 seconds that `__getitem__` takes 2600 are taken by transforms

So loading images is pretty fast, but transforming them is a lot longer. 
My approach with `load_images` isn't perfect and it also saves only 7% of training time, as you calculated, so maybe it is not very necessary to implement this optimization.

**Comment by AlekseySh** [comment]
Agree, sounds like too much changes for 7%
